Adds new algorithm for concave hull calculation - k nearest neighbours approach

Maximkaaa opened the pull request:
The short description of the algorithm is in the docs for the trait.

- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

lnicola commented on geo/src/algorithm/k_nearest_concave_hull.rs:
```diff
@@ -0,0 +1,348 @@
+    dataset.nearest_neighbor_iter(base_point).take(candidate_no as usize).collect()
+}
+
+fn sort_by_angle<'a, T>(points: &'a mut Vec<&Point<T>>, curr_point: &Point<T>, prev_angle: T)
```
Would it be possible to use a cross product here (signed area) instead of computing an angle?

Maximkaaa replied:
I've tried using dot-product method, but finally I've found even faster algorithm that works just fine.

michaelkirk commented:
I was hoping to hold off on reviewing this until this was addressed:

> The first one though looks like an obvious bug. There's actially a check in the code to see if the resulting hull contains all the input points. It probably has to do something with getting points out of the multipolygons. If you send me the geometries, I'll check it out. Any format would be fine. Json is good.

@dabreegster - have you sent the json?

dabreegster commented:
Sorry for the delay! These're serialized `Multipolygon<f64>`s.

The buggy case:
![buggy](https://user-images.githubusercontent.com/1664407/110380940-ac628500-800d-11eb-9219-9b493ca67624.png)
[buggy.json.txt](https://github.com/georust/geo/files/6104086/buggy.json.txt)

Some desired "shrinkwrapping" missing:
![shrink](https://user-images.githubusercontent.com/1664407/110381054-d5831580-800d-11eb-9365-88a8daed173f.png)
[large.json.txt](https://github.com/georust/geo/files/6104091/large.json.txt)

`k=10` for both, but I also tried 3.

Maximkaaa commented:
Thank you, @dabreegster, your application helped a lot in debugging this algorithm. There were a couple of bugs that I've found and fixed. After fixing, the result in case building concave polygons around the buildings looks like this:
![image](https://user-images.githubusercontent.com/6412426/114720310-977dbd80-9d40-11eb-8817-a73a6b75f2b5.png)
![image](https://user-images.githubusercontent.com/6412426/114720380-a5cbd980-9d40-11eb-90e7-c7b880727574.png)
![image](https://user-images.githubusercontent.com/6412426/114720421-ad8b7e00-9d40-11eb-9a9b-e38cd6d34c66.png)

There are still some issues that are not due to implementation but the limitation of the algorithm itself.

1. In some cases polygon looks like it could go more tightly around the objects:

![image](https://user-images.githubusercontent.com/6412426/114720783-fe02db80-9d40-11eb-8f1c-0d2e9f536b49.png)

The reason for this is that small house in the middle of the park. The algorithm is designed to always build a polygon with single contour. In case of clustered input objects, a rather larger K must be chosen to surround all input objects. One option is to let the method to return multipolygons, but that would break the logic for automatically adjust K, so it would have to be adjusted by hand in many cases.

2. If i switch to surround lanes instead of building, another issue appears:
![image](https://user-images.githubusercontent.com/6412426/114721694-da8c6080-9d41-11eb-8683-abc6c67ee996.png)
![image](https://user-images.githubusercontent.com/6412426/114721758-e7a94f80-9d41-11eb-84b3-9873d38836e5.png)

The reason for this is that the algorithm in its core only works with points. So to surround other shapes we just take all the vertexes and use them as an input. In the case when there are many little objects (like when we use buildings as input) it's not a problem as all vertexes are close to each other.

In the case when we have big long lines without any points, the algorithm may find some closer points to snap to, so not the whole polygon or line ends up to be inside of the hull. Although all the vertexes are actually inside of the hull.

One way to deal with this issue is to insert virtual vertexes on the long sides of polygons (or lines). But I failed to come up with a good generic way to choose how close I should insert those points.

Another options is to remove this treat from the Lines and Polygons altogether, to make it obvious that the algorithm with deal only with points.

What are your thoughts on this guys?

dabreegster commented:
> the result in case building concave polygons around the buildings looks like this:

These look much more "shrink-wrapped" -- thanks for iterating on this!

>  In case of clustered input objects, a rather larger K must be chosen to surround all input objects

This limitation makes sense; I don't think manually adjusting K would be feasible for my particular case, so I'm not opinionated about trying it.

> Another options is to remove this treat from the Lines and Polygons altogether, to make it obvious that the algorithm with deal only with points.

This limitation also makes sense, and yes, I would vote for only taking `Point`s in the method signature. I've been a bit confused by other geometric algorithms like generating Voronoi diagrams in the past, because from API documentation, it's not always clear that they don't work on entire polygons, and that you can't just feed in all of the polygon's boundary points.

michaelkirk commented:
> Another options is to remove this treat from the Lines and Polygons altogether, to make it obvious that the algorithm with deal only with points.

I'd prefer the approach of a signature which only accepts points for now.

To be correct, a hull needs to `Cover` its input. It might be surprising to people that the output could be incorrect in this way if the input is Polygons, especially since other implementations (e.g. PostGIS) do not behave this way. Note that our current concave_hull algorithm also has this limitation, and yet still accepts 1D and 2D geometries, but in hindsight I think this was a mistake.

For followup work, including a recipe in the documentation (and potentially helper methods) for interpolating points along a `line_string` and feeding that into `concave_hull` might be a good middle ground between making things not too tedious while also making clearer the limitations on this algorithm.

Maybe something like:
```
# note: this method doesn't currently exist
let subsampled_linestring = polygon.exterior().interpolated_with_max_distance(0.1);
let points = subsampled_linestring.0;
concave_hull(points)
```

(maybe relevant [`line_interpolate_point`](https://docs.rs/geo/0.18.0/geo/algorithm/line_interpolate_point/trait.LineInterpolatePoint.html))

And also, we should aspire to one day get a fancy concave hull algorithm that also works with 1D and 2D geometries. I'll open a separate issue for this...

michaelkirk commented on geo/src/algorithm/k_nearest_concave_hull.rs:
```diff
@@ -0,0 +1,393 @@
+    #[test_case(0., 0., -1., 0., PI)]
+    #[test_case(0., 0., 0., -1., PI / 2. * 3.)]
+    #[test_case(0., 0., -1., 1., PI / 4. * 3.)]
+    #[test_case(0., 0., -1., -1., PI / 4. * 5.)]
```
This is admittedly kind of nice, but I'd prefer to avoid adding an extra dependency for it.

(Also, curious if, as @lnicola mentioned, we could remove get_angle altogether)

michaelkirk commented on geo/src/algorithm/k_nearest_concave_hull.rs:
```diff
@@ -0,0 +1,393 @@
+/// The idea of the algorithm is simple:
+/// 1. Find a point on a future hull (e. g. a point with the smallest Y coordinate).
+/// 2. Find K nearest neighbours to the chosen point.
+/// 3. As the next point on the hull choose one of the nearest points, that would make the largest
```
```suggestion
/// 3. As the next point on the hull chose one of the nearest points, that would make the largest
```

michaelkirk commented on geo/src/algorithm/k_nearest_concave_hull.rs:
```diff
@@ -0,0 +1,393 @@
+/// In cases when the hull cannot be calculated for the given K, a larger value is chosen and
+/// calculation starts from the beginning.
+/// 
+/// In the worst case scenario, when no K can be found to build a correct hull, convex hull is
```
```suggestion
/// In the worst case scenario, when no K can be found to build a correct hull, the convex hull is
```

michaelkirk commented on geo/src/algorithm/k_nearest_concave_hull.rs:
```diff
@@ -0,0 +1,393 @@
+/// In the worst case scenario, when no K can be found to build a correct hull, convex hull is
+/// returned.
+/// 
+/// This algorithm is generally several times slower then the one used in 
```
```suggestion
/// This algorithm is generally several times slower than the one used in the
```

michaelkirk commented on geo/src/algorithm/k_nearest_concave_hull.rs:
```diff
@@ -0,0 +1,393 @@
+/// 
+/// The larger K is given to the algorithm, the more "smooth" the hull will generally be, but the
+/// longer calculation may take. If performance is not critical, K=3 is a safe value to set
+/// (lower values do not make sense for this algorithm). If K is equal or larger then the number of
```
```suggestion
/// (lower values do not make sense for this algorithm). If K is equal or larger than the number of
```

michaelkirk commented on geo/src/algorithm/k_nearest_concave_hull.rs:
```diff
@@ -0,0 +1,393 @@
+/// The larger K is given to the algorithm, the more "smooth" the hull will generally be, but the
+/// longer calculation may take. If performance is not critical, K=3 is a safe value to set
+/// (lower values do not make sense for this algorithm). If K is equal or larger then the number of
+/// input points, convex hull will be produced.
```
```suggestion
/// input points, the convex hull will be produced.
```

michaelkirk commented on geo/src/algorithm/k_nearest_concave_hull.rs:
```diff
@@ -0,0 +1,393 @@
+    where T: GeoFloat + RTreeNum
+{
+    let mut points: Vec<Coordinate<T>> = dataset.iter().map(|p| p.0).collect();
+    points.push(points[0]);
```
Upon further inspection... it seems like this method is only used when the original input has 3 or less points, so orderings here are all equivalent, but that's not obvious by reading this method.

So two alternatives I'd be OK with:
- do an `assert!(points.len() <= 3)` here to document this

or 

- delete this method and bail out earlier in the algorithm when we still have access to the initial point ordering rather than an RTree of those points.

Maximkaaa replied:
To use the second option we would have to change the algorithm signature to return Option<Polygon<T>>, so I prefer the first option.

Maximkaaa commented on geo/src/algorithm/k_nearest_concave_hull.rs:
```diff
@@ -0,0 +1,393 @@
+    concave_hull_inner(dataset, k)
+}
+
+const DELTA: f32 = 0.000000001;
```
I've changed the logic in points_equal() method and added a comment about this constant.

Maximkaaa commented:
Hi, @michaelkirk . Couldn't return to this earlier because of lack of time. I've fixed all the issues, mentioned in the review and updated it to the current master. Please, check if everything fine.

michaelkirk commented:
Hi @Maximkaaa. I made some time to review and QA the latest version, and things look good to me. Sorry for the slow turn around time!

I'll plan to merge this tomorrow unless I hear from someone else that they'd like to review it.

Maximkaaa commented:
@michaelkirk NP, totally understand you. It took me a month to make the fixes.

michaelkirk commented:
Squashed, formatted, and merged in https://github.com/georust/geo/commit/2c4261a4d413e18f3dfbcbc83ad551ae53e775cb
